chia/simulator/setup_services.py

```diff
@@ -441,57 +441,67 @@ async def setup_introducer(bt: BlockTools, port: int) -> AsyncGenerator[Service[
 
 
 @asynccontextmanager
-async def setup_vdf_client(bt: BlockTools, self_hostname: str, port: int) -> AsyncGenerator[asyncio.Task[Any], None]:
-    lock = asyncio.Lock()
+async def setup_vdf_client(bt: BlockTools, self_hostname: str, port: int) -> AsyncIterator[None]:
+    find_vdf_client()  # raises FileNotFoundError if not found
+    process_mgr = VDFClientProcessMgr(asyncio.Lock(), False, [])
     vdf_task_1 = asyncio.create_task(
-        spawn_process(self_hostname, port, 1, lock, prefer_ipv6=bt.config.get("prefer_ipv6", False))
+        spawn_process(self_hostname, port, 1, process_mgr, prefer_ipv6=bt.config.get("prefer_ipv6", False)),
+        name="vdf_client_1",
     )
 
     async def stop(
         signal_: signal.Signals,
         stack_frame: Optional[FrameType],
         loop: asyncio.AbstractEventLoop,
     ) -> None:
-        await kill_processes(lock)
+        await process_mgr.kill_processes()
 
     async with SignalHandlers.manage() as signal_handlers:
         signal_handlers.setup_async_signal_handler(handler=stop)
         try:
-            yield vdf_task_1
+            yield
         finally:
-            await kill_processes(lock)
+            await process_mgr.kill_processes()
+            vdf_task_1.cancel()
+            await asyncio.gather(vdf_task_1, return_exceptions=True)
```

Comment:
Yes, I originally just did `await` and while this works on macos fine, it did not on linux. It resulted in a hang as the await never returned - however, it's possible it would work now that I properly await the process ending after the kill. So I then tried a `cancel` and `await` and this ended up raising CancelledError in the await call - but I know it's Cancelled as I cancelled it so this just seemed less code than another try handler.
Note, I also just tried `cancel` but sometimes the process ends before the cancel is processed resulting in teardown errors about pending tasks. Since this is all teardown code, it seems consuming any exceptions while waiting for the task to finish after cancelling seemed reasonable to me

Reply:
1) I think most exceptions should be reported, we don't want silent failures we don't know about on shutdown.
2) The cancellation might not be yours...  but we won't deal with that in asyncio, at least not for now.
3) (actually actionable) I don't think that `asyncio.gather()` is a good way to express 'consume exceptions from a single call'.  I'd definitely rather see a `try`/`except`.  And I don't think using `asyncio.gather()` should have fixed anything vs. a direct `await`.